OMCompiler/Compiler/NBackEnd/Util/NBSlice.mo

```diff
@@ -332,7 +332,7 @@ public
 
     // sort
     for i in 1:arrayLength(indices) loop
-      indices[i] := List.sort(List.unique(indices[i]), intLt);
+      indices[i] := List.sort(UnorderedSet.unique_list(indices[i], Util.id, intEq), intLt);
     end for;
```

Comment:
My first thought was that this would be more counter-intuitive and slower than the sort-then-unique approach because in this case we first have to create a whole `UnorderedSet`, add all the elements from the list, then create a list again from the Unordered set. However, this is a bit faster. 

Here are the three options:

```mo
1. indices[i] := List.sort(List.unique(indices[i]), intLt);
2. indices[i] := List.sortedUnique(List.sort(indices[i], intLt), intEq);
3. indices[i] := List.sort(UnorderedSet.unique_list(indices[i], Util.id, intEq), intLt);
```

This is my interpretation

1. The original, is slow because it is using a `O(N^2)` unique algorithm on a big list. 
2. is asymptotically faster because it will first sort the list (hopefully in `O(N log(N))`) and then uses an `O(N)` unique algorithm (the list is now sorted and we only need to compare consecutive elements). So the whole thing is `O(N log(N))` at best.
3. is faster than 2. but only because of the structure of the input or the implementation. Even if we assume the insertion of all elements into the new `UnorderedSet` is done in the average O(1) time then the whole thing is O(N). Plus the creation of the new unique list from the `UnorderedSet` is also O(N). Then we have the sorting which is `O(N log(N))`. So the whole thing is `O(N log(N))` at best.


As it turns out 3. is the fastest for the `OneDHeatTransferTI_Modelica` on my machine. This makes me suspect that these lists are made mostly of duplicate elements. If we have a big list which is mostly made up of duplicate elements, then 3. might be faster than 1. because it has to sort a much smaller list. 

That said, if the data is supposed to be unique then maybe it is better to use UnorderedSet all over instead of representing it as a list. If it needs to be sorted we can use a Set (do we have one btw?)

Reply:
The `UnorderedSet` case could probably be improved further by setting the initial bucket count since we know how many elements we're going to add, which would avoid having to resize the set multiple times. I.e. something like:
```Modelica
marked_eqns_set = UnorderedSet.new(Util.id, intEq, Util.nextPrime(sum(listLength(l) for l in marked_eqns_set)));
```

Reply:
> As it turns out 3. is the fastest for the `OneDHeatTransferTI_Modelica` on my machine. This makes me suspect that these lists are made mostly of duplicate elements.

That's what I thought too. I also don't know what exactly happens in this code but I suspect that duplicates scale with the size of arrays or something like that.

> That said, if the data is supposed to be unique then maybe it is better to use UnorderedSet all over instead of representing it as a list.

The collection could be done with the set directly, yes. That might not be much faster, but maybe more readable...

Reply:
It seems setting the bucket size doesn't really matter here, because it's the code mentioned here using `UnorderedSet.unique_list` that's actually taking up all the time. And that's already using `UnorderedSet.fromList` internally which does set an appropriate bucket size.

However, I noticed that sorting the list afterwards accounts for a significant part of the time. I thought using an AvlSet instead might be faster since we can skip the sorting then, but it turned out to be slightly faster. The issue is probably that it allocates a lot more memory.

But I also tried implementing it as an external function using `std::sort` + `std::unique`, and even with the overhead of having to convert from a MetaModelica list to an `std::vector` and back it's still significantly faster:
| | Total time | Total allocations |
| ------------- | ------------- |------------- |
| List.sort + UnorderedSet  | 33.54s | 41.33 GB | 
| std::sort + std::unique  | 21.64s | 26.65 GB |
| AvlSetInt | ~35s | ~50 GB |

I can make a PR for my implementation if you want.

Reply:
How do these variants scale with N?

> I can make a PR for my implementation if you want.

It might be worth replacing `UnorderedSet.unique_list` altogether by your implementation then.

Reply:
> How do these variants scale with N?

O(N*log(N))
> > I can make a PR for my implementation if you want.
> 
> It might be worth replacing `UnorderedSet.unique_list` altogether by your implementation then.

My implementation is specialized for `list<Integer>` since it's an external C++ function, while `UnorderedSet.unique_list` work for any hashable type.

Reply:
I found `List.countingSort` which claims to be `O(N+n)` where `N` is the largest integer and `n` is the list size. That function could be modified to give a unique sorted list in one go.

Reply:
> The issue is probably that it allocates a lot more memory.

This reminds me that we have the problem that the module allocates quite a lot of memory. I will make a table for it but what I noticed was 3. used a bit less memory compared to 2. Maybe that accounts for the slight performance improvement between the two. And also add up to the total time in both cases. 

In any case though the whole module allocates a lot more memory than one would expect.

Reply:
Okay, in the meantime I'll merge this and talk with @kabdelhak about this next week. Maybe the whole idea behind `getDependentCrefIndicesPseudoArray` can be modified.

Reply:
> I found `List.countingSort` which claims to be `O(N+n)` where `N` is the largest integer and `n` is the list size. That function could be modified to give a unique sorted list in one go. We could even give an array as input so there would be fewer allocations, right?

Sure, if you know that N will be approximately equal to n then you can use a bucket sort which would be the fastest way to do it. Whether you give an array or list as input doesn't really matter much since it just going to be iterated through, but using arrays in general will be faster.

> > The issue is probably that it allocates a lot more memory.
> 
> This reminds me that we have the problem that the module allocates quite a lot of memory. I will make a table for it but what I noticed was 3. used a bit less memory compared to 2. Maybe that accounts for the slight performance improvement between the two. And also add up to the total time in both cases.
> 
> In any case though the whole module allocates a lot more memory than one would expect.

Yeah, I observed that it's sorting what looks like the same huge lists over and over thousands of times in a row, which I think is the reason for the memory usage. But I don't know what the algorithm is actually doing.